### ISSUE TITLE
Cheevos config only saves the first 31 characters of the password

### ISSUE THREAD

**Issue by computertoucher** [libretro/RetroArch#11169]
## Description

When inputting a password for a RetroAchievements account, only the first 31 characters of the password is saved in the configuration file. This also happens if one edits the configuration file manually to include the complete password -- when a config is saved new or overwritten, the field contains only 31 characters.

### Expected behavior

The complete password is stored without truncation.

### Actual behavior

RetroArch truncates anything above 31 characters in the RetroAchievements password field.

### Steps to reproduce the bug

1. Input username and password for Achievements
2. Save your configuration either manually as a new file or overwrite the old.
3. Exit RetroArch.
4. Open the .cfg file, notice the cheevos_password configuration field is truncated to 31 characters.
5. Paste the actual long password into the field
6. Open RetroArch again, load the appropriate config
7. Save that config again. The password is truncated again.

### Bisect Results

Since 1.8.5 at least, also on 1.9.0 after upgrading.

### Version/Commit
You can find this information under Information/System Information

- RetroArch: 1.8.5 and 1.9.0

I think this is the culprit, or at least has something to do with the issue -- but am not skilled enough to know for sure:

https://github.com/libretro/RetroArch/blob/b1723f7cd064cb402197b0e34d143347713064a9/configuration.h#L672

### Environment information

- OS: Windows 10 build 2004

**Comment by Nargash** [comment]
Hi @computertoucher, I was thinking of trying to fix this. Would 63 characters be a reasonable password limit instead? I believe cheevos.c would also need to be updated if the limit was changed higher than that. Ideally the password wouldn't be stored in plain text but fixing that is probably beyond my abilities. Thanks!

**Comment by Jamiras** [comment]
it's only stored in plaintext until the first time the user logs in, at which point it cleared out and a token is stored in `cheevos_token`.

**Comment by computertoucher** [comment]
Hi @Nargash I'll start with a tl;dr: For practical reasons, since this is just to interface with another service and would be restricted by their requirements, I'd check with the RetroAchievements devs to see what their maximum length is. Probably 128 or 255 characters. And now here's a rant:

Personally I think password fields should not be practically restricted by length, and regularly use passphrases that exceed 64 characters (like 8-12 words).

So if you're desiging a new database structure for your app, if an arbitrary length absolutely must be enforced for some reason, a standard VARCHAR can hold 255 characters and should (in this day and age) be enough. I'd prefer it to be a TEXT or LONGTEXT so one can use password files of up to 6KiB or even 4MiB but for now that is pretty impractical.

Not allowing passphrases up to 255 characters though, in these days that's just stupid.

**Comment by Jamiras** [comment]
As far as I know, the password is not length restricted by the service. Once received on the server, it's encrypted with salt, and compared to the fixed-length hash stored in the database. The password field when creating an account is currently unbounded.

As I've commented above, once the service authenticates the user, a token (also fixed length) is returned and used for all future interactions. The current limitation is purely in the way RetroArch manages the password until the login is performed as noted by the code you've highlighted.

**Comment by computertoucher** [comment]
@Jamiras Yes, and that is what I propose fixing? Since it's, like, impossible to log in that first time -- to get that token -- with a password of more than 31 characters.

Either that CHAR allocation must increase in size, or the cheevos login / token fetch routine must be altered to e.g. query the service and then store the token if successful -- without temporarily storing the password in the config file.

Expecting users to change their cheevos password to something shorter when the fault is in RetroArch is not the right way to do it, though it's that's currently the only way to fix it if you're a regular user.

**Comment by Nargash** [comment]
I was hoping to find an easy issue to fix! If the password length is truly unlimited, some large value (like 255 that was suggested) would probably have to be settled on. One reason is Retroarch seems to send the password in a URL in [rc_url_login_with_password](https://github.com/libretro/RetroArch/blob/a8eaf82e065489504aea8ee1766c4bf677442576/deps/rcheevos/src/rurl/url.c#L168). Many web browsers have an upper limit of ~2048 characters.

Interestingly I've also observed the password not getting cleared and saving as a token in retroarch.cfg upon first login, whenever core overrides are loaded - but that is another story to investigate!

**Comment by Hoff16** [comment]
This came up in a search I did and can certainly be closed.
